sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -202,19 +203,58 @@ class Analyzer(
     Batch("Subquery", Once,
       UpdateOuterReferences),
     Batch("Cleanup", fixedPoint,
-      CleanupAliases)
+      CleanupAliases),
+    Batch("Recursion", Once,
+      CheckRecursionConstraints)
   )
 
+  object ResolveRecursiveReferneces extends Rule[LogicalPlan] {
```

Comment:
Hmm, I don't think so. Resolving a `RecursiveTable` has 2 phases. Resolving the anchor terms first which let us resolve the table's `UnresolvedRecursiveReference`s to `RecursiveReference`s (in rule`ResolveRecursiveReferneces`) and so in the second phase the `RecursiveTable` can be resolved by the usual rules.
At the point when `CTESubstitution` runs we can't insert `RecursiveReference` because we don't know the attributes coming from the anchors.

Reply:
I just realized that this can be a good catch. Let me look into this...

Reply:
I remember now. The thing is that when we have nested recursion and the outer recursive reference is in the anchor of the inner recursion like this:
```
-- recursion nested into recursion 2
WITH RECURSIVE t AS (
  WITH RECURSIVE s AS (
    SELECT j, 1 AS i FROM t
    UNION ALL
    SELECT j, i + 1 FROM s WHERE i < 3
  )
  VALUES (1) as t(j)
  UNION ALL
  SELECT j + 1 FROM s WHERE j < 3
)
SELECT * FROM t;
```
analyzing the inner will not resolve it's anchor and so we can't place resolved `RecursiveReference s` immediately into the plan.